File: sketches/ellie-03/sketch.fs
uniform float uTime;
uniform vec2 uResolution;

uniform vec3 uColor1;
uniform vec3 uColor2;
uniform vec3 uColor3;
uniform vec3 uColor4;

varying vec2 vUv;

float random (vec2 st) {
	return fract(sin(dot(st.xy, vec2(12.9898,78.233)) * 43758.5453123));
}

void main() {
	vec2 st = fract(vUv * 20.0);
	vec2 grid = floor(vUv * 20.0);

    vec2 normalizedGrid = grid / uResolution.xy * 0.001;

    float randomValue = random(normalizedGrid * cos(-uTime * 0.0000001) * mod(-uTime * 0.000001, 10000.0)) * random(normalizedGrid * sin(uTime * 0.00000001) * 1000.0); // Pass the normalized grid coordinates to the random function

	float yPos = 1.0 - vUv.y;
	float yPos2 = 1.0 - st.y;

	vec3 gradient = mix(uColor1, uColor2, smoothstep(0.0, 0.26, yPos));
	gradient = mix(gradient, uColor3, smoothstep(0.26, 0.445, yPos));
	gradient = mix(gradient, uColor4, smoothstep(0.445, 1.0, yPos));

	vec3 gradient2 = mix(uColor1, uColor2, smoothstep(0.0, 0.26, yPos2));
	gradient2 = mix(gradient2, uColor3, smoothstep(0.26, 0.445, yPos2));
	gradient2 = mix(gradient2, uColor4, smoothstep(0.445, 1.0, yPos2));

    vec3 color = randomValue > 0.5 ? vec3(1.0) : vec3(0.0);

	vec3 finalColor = mix(gradient, gradient2, color);

    gl_FragColor = vec4(finalColor, 1.0);

    #include <colorspace_fragment>
}

File: sketches/ellie-03/sketch.js
import * as THREE from 'three';
import fragmentShader from './sketch.fs';

export let props = {
	color1: {
		value: '#C2C3DA',
		onChange: ({ value }) => {
			uniforms.uColor1.value.set(value)
		}
	},
	color2: {
		value: '#FFC165',
		onChange: ({ value }) => {
			uniforms.uColor2.value.set(value)
		}
	},
	color3: {
		value: '#FF3A3A',
		onChange: ({ value }) => {
			uniforms.uColor3.value.set(value)
		}
	},
	color4: {
		value: '#0F144E',
		onChange: ({ value }) => {
			uniforms.uColor4.value.set(value)
		}
	}
}

let camera;
let uniforms = {
	uResolution: { value: new THREE.Vector2() },
	uTime: { value: 0 },
	uColor1: { value: new THREE.Color(props.color1.value) },
	uColor2: { value: new THREE.Color(props.color2.value) },
	uColor3: { value: new THREE.Color(props.color3.value) },
	uColor4: { value: new THREE.Color(props.color4.value) }
};

/**
 * @param {object} params
 * @param {HTMLCanvasElement} params.canvas
 * @param {THREE.Renderer} params.renderer
 * @param {THREE.Scene} params.scene
 * @param {number} params.width
 * @param {number} params.height
 * @param {number} params.pixelRatio
 */
export let init = ({ scene, width, height }) => {
	camera = new THREE.OrthographicCamera(1, 1, 1, 1, 1, 1000);

	let geometry = new THREE.BufferGeometry();
	geometry.setAttribute(
		'position',
		new THREE.Float32BufferAttribute([-1, 3, 0, -1, -1, 0, 3, -1, 0], 3),
	);
	geometry.setAttribute(
		'uv',
		new THREE.Float32BufferAttribute([0, 2, 0, 0, 2, 0], 2),
	);

	let mesh = new THREE.Mesh(
		geometry,
		new THREE.ShaderMaterial({
			vertexShader: `
        varying vec2 vUv;

        void main() {
            vUv = uv;
            gl_Position = vec4(position, 1.);
        }
        `,
			fragmentShader,
			uniforms,
		}),
	);

	scene.add(mesh);
};

/**
 * @param {object} params
 * @param {HTMLCanvasElement} params.canvas
 * @param {THREE.Renderer} params.renderer
 * @param {THREE.Scene} params.scene
 * @param {number} params.width
 * @param {number} params.height
 * @param {number} params.pixelRatio
 * @param {number} params.time
 * @param {number} params.deltaTime
 * @param {number} params.frame
 * @param {number} params.playhead
 * @param {number} params.playcount
 */
export let update = ({ renderer, scene, time, deltaTime }) => {
	uniforms.uTime.value = time;

	renderer.render(scene, camera);
};

/**
 * @param {object} params
 * @param {HTMLCanvasElement} params.canvas
 * @param {THREE.Renderer} params.renderer
 * @param {THREE.Scene} params.scene
 * @param {number} params.width
 * @param {number} params.height
 * @param {number} params.pixelRatio
 */
export let resize = ({ width, height }) => {
	uniforms.uResolution.value.x = width;
	uniforms.uResolution.value.y = height;

	camera.left = -width * 0.5;
	camera.right = width * 0.5;
	camera.top = height * 0.5;
	camera.bottom = -height * 0.5;

	camera.updateProjectionMatrix();
};

export let rendering = 'three';

export let buildConfig = {
	gui: {
		output: true
	}
}